tools/clang/lib/SPIRV/EmitSPIRVAction.cpp

```diff
@@ -379,13 +441,26 @@ class SPIRVEmitter : public ASTConsumer {
 
     TranslationUnitDecl *tu = context.getTranslationUnitDecl();
 
-    // Process all top level Decls.
+    // A queue of functions we need to translate.
+    std::deque<FunctionDecl *> workQueue;
+
+    // The entry function is the seed of the queue.
     for (auto *decl : tu->decls()) {
-      doDecl(decl);
+      if (auto *funcDecl = dyn_cast<FunctionDecl>(decl)) {
+        if (funcDecl->getName() == entryFunctionName) {
+          workQueue.push_back(funcDecl);
+        }
+      }
+    }
+
+    // Translate all functions reachable from the entry function.
+    while (!workQueue.empty()) {
+      doFunctionDecl(workQueue.front());
```

Comment:
We should enlarge the workQueue when seeing a function call. I will add a TODO for that.